src/thingset_common.c

```diff
@@ -76,7 +122,7 @@ int thingset_common_serialize_record(struct thingset_context *ts,
     }
 
     /* record item definitions are expected to start behind record data object */
-    const struct thingset_data_object *item = object + 1;
+    const struct thingset_data_object *item = thingset_get_object_by_id(ts, object->id) + 1;
```

Comment:
When I was writing this, I think the assumption that the record item definition was the next item in the data object array was not true. Perhaps it is again now. I will test without this change.

Reply:
As discussed, this is needed because `object` is no longer necessarily a member of the `data_objects` array (it may have been created by a call to `thingset_common_prepare_record_element` further up the stack). So we get the "real" `thingset_data_object` (i.e. the one in the array) and then index using that.